cpp/velox/shuffle/VeloxSortShuffleWriter.cc

```diff
@@ -0,0 +1,315 @@
+    auto pidArr = getFirstColumn(*(pidBatch->getRowVector()));
+    RETURN_NOT_OK(partitioner_->compute(pidArr, pidBatch->numRows(), row2Partition_, partition2RowCount_));
+
+    auto rvBatch = VeloxColumnarBatch::from(veloxPool_.get(), batches[1]);
+    return rvBatch->getFlattenedRowVector();
+  } else {
+    auto veloxColumnBatch = VeloxColumnarBatch::from(veloxPool_.get(), cb);
+    VELOX_CHECK_NOT_NULL(veloxColumnBatch);
+    auto rv = veloxColumnBatch->getFlattenedRowVector();
+    if (partitioner_->hasPid()) {
+      auto pidArr = getFirstColumn(*rv);
+      RETURN_NOT_OK(partitioner_->compute(pidArr, rv->size(), row2Partition_, partition2RowCount_));
+      return getStrippedRowVector(*rv);
+    } else {
+      RETURN_NOT_OK(partitioner_->compute(nullptr, rv->size(), row2Partition_, partition2RowCount_));
+      return rv;
+    }
+  }
+}
+
+arrow::Status VeloxSortShuffleWriter::insert(const facebook::velox::RowVectorPtr& vector, int64_t memLimit) {
+  ScopedTimer timer(&c2rTime_);
+  auto inputRows = vector->size();
+  VELOX_DCHECK_GT(inputRows, 0);
+
+  facebook::velox::row::CompactRow row(vector);
+
+  if (!fixedRowSize_) {
+    rowSizes_.resize(inputRows + 1);
+    rowSizes_[0] = 0;
+    for (auto i = 0; i < inputRows; ++i) {
+      rowSizes_[i + 1] = rowSizes_[i] + row.rowSize(i);
+    }
+  }
+
+  uint32_t rowOffset = 0;
+  while (rowOffset < inputRows) {
+    auto remainingRows = inputRows - rowOffset;
+    auto rows = maxRowsToInsert(rowOffset, remainingRows);
+    if (rows == 0) {
+      auto minSizeRequired = fixedRowSize_ ? fixedRowSize_.value() : rowSizes_[rowOffset + 1] - rowSizes_[rowOffset];
+      acquireNewBuffer(memLimit, minSizeRequired);
+      rows = maxRowsToInsert(rowOffset, remainingRows);
+      ARROW_RETURN_IF(
+          rows == 0, arrow::Status::Invalid("Failed to insert rows. Remaining rows: " + std::to_string(remainingRows)));
+    }
+    RETURN_NOT_OK(maybeSpill(rows));
+    insertRows(row, rowOffset, rows);
+    rowOffset += rows;
+  }
+  return arrow::Status::OK();
+}
+
+void VeloxSortShuffleWriter::insertRows(facebook::velox::row::CompactRow& row, uint32_t offset, uint32_t rows) {
+  // Allocate newArray can trigger spill.
+  growArrayIfNecessary(rows);
+  for (auto i = offset; i < offset + rows; ++i) {
+    auto size = row.serialize(i, currentPage_ + pageCursor_);
+    array_[offset_++] = {toCompactRowId(row2Partition_[i], pageNumber_, pageCursor_), size};
+    pageCursor_ += size;
+  }
+}
+
+arrow::Status VeloxSortShuffleWriter::maybeSpill(int32_t nextRows) {
+  if ((uint64_t)offset_ + nextRows > std::numeric_limits<uint32_t>::max()) {
+    EvictGuard evictGuard{evictState_};
+    RETURN_NOT_OK(evictAllPartitions());
+  }
+  return arrow::Status::OK();
+}
+
+arrow::Status VeloxSortShuffleWriter::evictAllPartitions() {
+  {
+    ScopedTimer timer(&sortTime_);
+    // TODO: Add radix sort to align with Spark.
+    std::sort(array_.begin(), array_.begin() + offset_);
```

Comment:
can we use the prefix sort in Velox here?

Reply:
Seems like we cannot benefit from prefix sort here. By following Spark's design, `array_` stores a compacted id (uint64_t) to the row https://github.com/apache/incubator-gluten/pull/6475/files#diff-ff82d0ed4685583538881cabfe93ed92f202dd182492a66937fba0ccd5f7a22fR33-R36 Only the `array_` is sorted here, not the entire row.